Default value changes in DeepTauId module and puppiweight functions

lwezenbe opened the pull request:
Changes in this PR:
- Remove the default_value constant from all deepTau v2 code
- Changed default value for variables related to number of hits to 0 instead of -999
- Change the puppiweight to have an additional argument to define its return value for AOD. This allows the puppiweights to be defined separately for different blocks. Additionally split between weights for inner and outer blocks. All values have been extracted from the training.
- Removed default_value argument from certain candfunctions and instead use conditionals where needed in the places these functions are called.

Tests:
- Reran config file with update from deepTauAtHLT.py and everything ran smoothly. Running RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py also ran fine. Upon brief inspection, no strange behaviour observed.

mbluj commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -630,9 +630,21 @@ namespace {
-    auto getPuppiWeight(const reco::PFCandidate& cand) { return 0.9907f; }
+    auto getPuppiWeight(const reco::PFCandidate& cand) { 
+      if (std::abs(cand.pdgId()) == 11) return 0.9787f;
+      else if (std::abs(cand.pdgId()) == 13) return .9907f;
```
Please add 0 before decimal point (here and other new lines where it is missing)
Please also confirm that you have run
```
scram b code-check
scram b code-format
```
as I am not sure if current formatting agrees with CMS rules.
BTW, literally, `else` statements are not mandatory here because of `return` executed when condition is fulfilled. Maybe `case` is nicer construction here? With `case`  `std::abs(cand.pdgId())` will be evaluated just once which is also a plus.

lwezenbe replied:
Done

mbluj replied:
Thanks.
Could you please put zero before decimal point consistently for all default values here and in the `getPuppiWeightNoLep` method? It can be done together with other changes requested by Konstantin.

mbluj replied:
Cosmetics: maybe it makes also sense to order pdgIds in the same way in both methods, e.g. (11,13) 22, 211, 130.
BTW, is it understood why default weights in both functions are identical for charged hadrons (211), similar for photons (22) and quite different for neutral hadrons (130)?

kandrosov replied:
> BTW, is it understood why default weights in both functions are identical for charged hadrons (211), similar for photons (22) and quite different for neutral hadrons (130)?

@lwezenbe what are the origins of the default values for cases where information is missing? Did you take them from Maria Teresa's code? I remember that the values she used were preliminary and maybe for the central CMSSW integration we should re-evaluate them more accurately. Also, the mean values must be different for inner and outer cells (this was not implemented in Maria Teresa's version).

lwezenbe replied:
Actually, I think I probably added them myself when changing some defaults. In Maria-Teresa's code, she had only AOD, so she could just set the value in the "get(...) = " lines themselves, as here for example:
https://github.com/mgrippo/cmssw/blob/deepTau_hlt_dev/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L1319
So if we need to re-evaluate even more accurately, how should we go about it for now?

kandrosov replied:
Best way is to extract values directly from DeepTau v2 training root-tuples (currently being copied to /eos/cms/store/group/phys_tau/TauML/DeepTau_v2/tuples-v2-training-v2-t1-root/training/).
E.g.
```python
file = '/eos/cms/store/group/phys_tau/TauML/DeepTau_v2/tuples-v2-training-v2-t1-root/training/part_0.root'
tree = 'inner_cells'
val_flag = 'pfCand_ele_valid'
var = 'pfCand_ele_puppiWeight'

df = ROOT.RDataFrame(tree, file)
x = df.Filter('{} > 0'.format(val_flag)).Mean(var)
print('E[{}] = {}'.format(var, x.GetValue()))
```

mbluj commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -630,9 +630,21 @@ namespace {
     auto getPuppiWeight(const pat::PackedCandidate& cand) { return cand.puppiWeight(); }
-    auto getPuppiWeightNoLep(const reco::PFCandidate& cand) { return 0.8858f; }
+    auto getPuppiWeightNoLep(const reco::PFCandidate& cand) { 
+      if (std::abs(cand.pdgId()) == 211) return 0.7614f; 
```
Same comments as to `getPuppiWeight` method.

mbluj reviewed (commented):
@lwezenbe thank you for this PR. A few comments inline.

swozniewski commented:
Thanks for the quick update! Does the deepTauID application fall back on the changed defaults with pat.objects? Do we hence expect changes in the DQM checks?

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -1547,6 +1559,8 @@ class DeepTauId : public deep_tau::DeepTauBase {
                 << "invalid prediction = " << pred << " for tau_index = " << tau_index << ", pred_index = " << k;
           predictions.matrix<float>()(tau_index, k) = pred;
         }
+      } else {
+        continue;
```
what is the purpose of this?

mbluj replied:
I think that the intention was to ignore taus not passing discriminants, but for such a case anyway noting happens in the loop below, so I tend to agree that it is not needed. Is it correct?

lwezenbe replied:
Indeed, it was. I copied these changes from a debugging repo and for these two lines I think I misread the use and need for them because when examining a bit closer now, it indeed seems like it is not needed. I will remove them.

lwezenbe commented:
> Thanks for the quick update! Does the deepTauID application fall back on the changed defaults with pat.objects? Do we hence expect changes in the DQM checks?

It does not. As you can see here, it does not use the default value when calling the same function for pat objects:
 https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_2_X_tau-pog_deepTauForHLT_phase2/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L642
Here it just returns the same value as in the original code before the AOD implementation was done.

kandrosov commented:
Why do we have `default_value = -999`? As far as I know, there are no inputs that have this value as the default one...

https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_2_X_tau-pog_deepTauForHLT_phase2/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L1108

lwezenbe commented:
> Why do we have `default_value = -999`? As far as I know, there are no inputs that have this value as the default one...
> 
> https://github.com/cms-tau-pog/cmssw/blob/CMSSW_11_2_X_tau-pog_deepTauForHLT_phase2/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L1108

I am not sure why the value -999 was originally put there, but I just kept this default_value from the original code, as seen here in the CMSSW_11_1_X version:
https://github.com/cms-sw/cmssw/blob/CMSSW_11_1_X/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L810
As far as I can see, it was used there for every case where no pat::candidate object was found

lwezenbe commented:
I changed the if-else statements in the puppiweight functions to switch expressions, removed the extra lines now and have run:
`scram b code-check`
`scram b code-format`
I also reran the configuration file to test and everything still ran smoothly

swozniewski commented:
Ok, thanks a lot! No further questions from my side.

kandrosov commented:
> I am not sure why the value -999 was originally put there, but I just kept this default_value from the original code, as seen here in the CMSSW_11_1_X version:

Ok, I see. So it was used in DeepTau v1 as the default value, I've already forgot it :) 
The only line where it was used for DeepTau v2 was here [1] to set `tau_dz_sig` in the same way as was done during the training. So it is also kind of "compatibility workaround".

Bottom line, don't use `default_value` anywhere else. And for the next code cleaning, we should remove it together with all DeepTau v1 related code.

[1] https://github.com/cms-sw/cmssw/blob/CMSSW_11_1_X/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L1236

lwezenbe commented:
> > I am not sure why the value -999 was originally put there, but I just kept this default_value from the original code, as seen here in the CMSSW_11_1_X version:
> 
> Ok, I see. So it was used in DeepTau v1 as the default value, I've already forgot it :)
> The only line where it was used for DeepTau v2 was here [1] to set `tau_dz_sig` in the same way as was done during the training. So it is also kind of "compatibility workaround".
> 
> Bottom line, don't use `default_value` anywhere else. And for the next code cleaning, we should remove it together with all DeepTau v1 related code.
> 
> [1] https://github.com/cms-sw/cmssw/blob/CMSSW_11_1_X/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L1236

Yes, I was just coming to that conclusion as well. I must have gotten that mixed up while implementing the AOD part. I can change the default values in all the places where I used it erroneously, I will try to do it today although I am not sure how much time I'll have.

lwezenbe commented:
I was just going through the module again to remove any default_values that a should not be used. I noticed however that while there were a few that should have been set to 0, most of them (when used in v2 at least) occur in blocks like this:
https://github.com/cms-tau-pog/cmssw/blob/b5f21e45ad128af66e2a1d7ed524c7773ad28845/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L2295-L2308
Here the condition for using the default value is if hasTrackDetails if false, which per definition can not be the case when the function is called. The reason default values are implemented in those functions are for cases like this:
https://github.com/cms-tau-pog/cmssw/blob/b5f21e45ad128af66e2a1d7ed524c7773ad28845/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L1951
Where this was not checked first.
Is there a preference to keep it as "default_value" in the cases where it is never used or to change these cases to a value like 0?

kandrosov commented:
The original behaviour is to fill `pfCand_muon_dz_sig` with 0 if ` pfCands.at(index_pf_muon).hasTrackDetails()` is `false`. Same for other cases. Except `tau_dz_sig`, where `std::abs(leadChargedHadrCand->dz()) / -999.` is passed to `getValueNorm` (for the cases where `hasTrackDetails` is `false`), but this was an error during the training, so kept here to have bug-to-bug compatibility. I don't find any other exception (but it can be that I'm missing something).

IMO, adding new default value (even if this part of the code, in theory, shouldn't be reachable) will only add confusion.

lwezenbe commented:
Maybe it would be better to get the default_value out of the getTauDz functions and so on and use a conditional operator as Maria Teresa did here in those cases where a default of 0 is needed:
https://github.com/mgrippo/cmssw/blob/deepTau_hlt_dev/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L805
And then there does not need a default value set in the other blocks to avoid confusion.
My only worry is that if the code is ever changed or there is an edge case that we didnt think of where it can be reached, the code will not work. Unless we make the argument in the getTauDz function a default argument so we do not need to explicitely put potentially confusing "default_value" in one of those blocks I linked before to reduce the confusion.
What do you think?

kandrosov commented:
> Maybe it would be better to get the default_value out of the getTauDz functions and so on and use a conditional operator

Yes, I agree, it should be more transparent.

lwezenbe commented:
I have made a few changes:
- Removed the switch statement from the puppiweight functions and instead added an extra argument to the puppiweight functions that can be defined differently for each block (so it does not have to rely on potentially using a self-defined possibly erroneous default value of 0)
- Updated the puppiweights by retrieving them from the training as Konstantin commented. A split is now made between inner and outer cells as well
- Removed the default_value constant from all deeptau v2 related code. In the only spot where it should be used, I explicitely wrote -999. Everywhere else a default value is defined, it was taken from Maria-Teresa's code
- Removed the default_argument from some functions to reduce confusion and instead used conditionals where needed

swozniewski commented:
Thanks! I'll have a look. Should it be complete now?

swozniewski commented:
@lwezenbe please double-check the PR description if not already done. It would be helpful if I could copy it to the integration PR.

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -2847,4 +2841,4 @@ class DeepTauId : public deep_tau::DeepTauBase {
 
 #include "FWCore/Framework/interface/MakerMacros.h"
-DEFINE_FWK_MODULE(DeepTauId);
+DEFINE_FWK_MODULE(DeepTauId);
```
please undo this change

lwezenbe commented:
> @lwezenbe please double-check the PR description if not already done. It would be helpful if I could copy it to the integration PR.

I have adapted the PR description to the latest changes made. As far as testing goes, they were mainly tests to see if everything ran fine. A visual inspection showed no strange results, but I have not had the time to do more in-depth testing.
